allennlp/data/dataset_readers/coreference_resolution/conll.py

```diff
@@ -0,0 +1,254 @@
+                            cluster_id = int(segment[1:])
+                            document_state.coref_stacks[cluster_id].append(word_index)
+                    else:
+                        # The span for this id is ending, but didn't start at this word.
+                        # Retrieve the start index from the document state and
+                        # add the span to the clusters for this id.
+                        cluster_id = int(segment[:-1])
+                        start = document_state.coref_stacks[cluster_id].pop()
+                        document_state.clusters[cluster_id].append((start, word_index))
+
+
+class _DocumentState:
+
+    """
+    Represents the state of a document. Words are collected in
+    a sentence buffer, which are incrementally collected in a
+    list when sentences end, representing all tokens in the document.
+
+    Additionally, this class contains a per-id stacks to hold the start indices of
+    active spans (spans which we are inside of when processing a given word). Spans
+    with the same id can be nested, which is why we collect these opening spans
+    on a stack, e.g:
+
+     [Greg, the baker who referred to [himself]_ID1 as 'the bread man']_ID1
+
+    Once an active span is closed, the span is added to the cluster for the given id.
+    """
+
+    def __init__(self) -> None:
+        self.sentence_buffer: List[str] = []
+        self.sentences: List[List[str]] = []
+        self.num_total_words = 0
+        # Cluster id -> List of (start_index, end_index) spans.
+        self.clusters: DefaultDict[int, List[Tuple[int, int]]] = collections.defaultdict(list)
+        # Cluster id -> List of start_indices which are open for this id.
+        self.coref_stacks: DefaultDict[int, List[int]] = collections.defaultdict(list)
+
+    def assert_finalizable(self) -> None:
+
+        if self.sentence_buffer:
+            raise ConfigurationError("Processing was attempted on a document which had "
+                                     "incomplete sentences. Current sentence buffer "
+                                     "contains: {}".format(self.sentence_buffer))
+
+        if not self.sentences or self.num_total_words == 0:
+            raise ConfigurationError("Processing was attempted on a document with no sentences or words.")
+
+        if any(x for x in self.coref_stacks.values()):
+            raise ConfigurationError("Processing was attempted on a document which contains incomplete "
+                                     "(unclosed) spans within the scope of the document. Current ids "
+                                     "and start indices" " of unclosed spans: {}".format(self.coref_stacks))
+
+    def complete_sentence(self) -> None:
+        self.sentences.append(self.sentence_buffer)
+        self.sentence_buffer = []
+
+    def add_word(self, word) -> None:
+        self.sentence_buffer.append(word)
+        self.num_total_words += 1
+
+    def finalize(self) -> List[Tuple]:
```

Comment:
@kentonl - this method has stumped me a bit. Why do we ever need to merge clusters with different ids? Do we ever get exactly overlapping spans with different ids in the data?

Reply:
If the dataset were perfect, this would not be needed. Unfortunately, the same mention can appear in multiple clusters. This happens exactly once in the training set, once in the dev set, and never in the test set.

Reply:
I can imagine the deep sense of rage you must have felt when you discovered that. Thanks!